File: exercism/python/crypto-square/crypto_square.py
# https://exercism.io/my/solutions/8d8f2c7494514351862390ebf6b23271

from math import sqrt, floor
from textwrap import wrap


def dimensions(n):
    rows = round(sqrt(n))
    columns = rows if n == rows ** 2 else floor(sqrt(n)) + 1
    return rows, columns


def cipher_text(plain_text):
    normalized = ''.join([c for c in plain_text if c.isalnum()]).lower()
    if normalized == '':
        return ''

    rows, columns = dimensions(len(normalized))
    rectangle = wrap(normalized, columns)
    rectangle[-1] = rectangle[-1].ljust(columns, ' ')

    encoded = []
    for c in range(columns):
        encoded.append(''.join([rectangle[r][c] for r in range(rows)]))
    return ' '.join(encoded)

File: exercism/go/crypto-square/crypto_square.go
// https://exercism.io/my/solutions/7aa7eb5ea75141f7a7979c12b0e5c212

package cryptosquare

import (
	"math"
	"strings"
	"unicode"
)

func normalize(s string) string {
	alphanumToLower := func(r rune) rune {
		if !unicode.IsLetter(r) && !unicode.IsDigit(r) {
			return -1
		}
		return unicode.ToLower(r)
	}

	return strings.Map(alphanumToLower, s)
}

func dimensions(n int) (int, int) {
	height := int(math.Round(math.Sqrt(float64(n))))
	var width int
	if height*height == n {
		width = height
	} else {
		width = int(math.Floor(math.Sqrt(float64(n)))) + 1
	}
	return height, width
}

// Encode encodes a string using a cryptosquare
func Encode(s string) string {
	plain := []rune(normalize(s))
	n := len(plain)
	h, w := dimensions(n)
	square := make([][]rune, h)

	for r := 0; r < h; r++ {
		row := make([]rune, w)
		for c := 0; c < w; c++ {
			index := (r * w) + c
			if index >= n {
				row[c] = ' '
			} else {
				row[c] = plain[index]
			}
		}
		square[r] = row
	}

	var b strings.Builder
	for c := 0; c < w; c++ {
		for r := 0; r < h; r++ {
			b.WriteRune(square[r][c])
		}
		if c != w-1 {
			b.WriteRune(' ')
		}
	}
	return b.String()
}
